nginx/config.go

```diff
@@ -186,6 +194,7 @@ http {
 		ssl_protocols {{ $sslConfig.Protocols }};
 		{{ if ne $sslConfig.Ciphers "" }}ssl_ciphers {{ $sslConfig.Ciphers }};{{ end }}
 		ssl_prefer_server_ciphers on;
+		ssl_early_data {{ if ne $sslConfig.EarlyDataMethods "" }}on{{ else }}off{{ end }};
```

Comment:
So `ssl_early_data` is turned off if no methods are set here? Is that the behaviour we want as default. Seems like ssl_early_data makes everything more secure in TLSv1.3?

Reply:
Yes, that can be used to disable it. The default is "GET|HEAD|OPTIONS" which are the HTTP methods considered idempotent and safe. 

Early-Data actually makes things less secure, but speeds up initial connections from 3 to 2 roundtrips, making them as fast as session resumptions.

Early-Data allows to send the http request like GET / as part of the SSL handshake, but since the handshake is not completed yet, it can be used for replay attacks. 

If someone captures the tcp packets of a connection handshake with early data, those packets cannot be decoded, but can be replayed.

This is the reason that only safe and idempotent HTTP methods are allowed by default, since replaying them should be harmless.

Of course if application developers ignore HTTP conventions and use GET requests to delete records or initiate payments, they are in trouble. 

This is why the router adds the standard "Early-Data: 1" request header and apps can reply with the "425 Too Early" HTTP status code, which would cause the client to retry the request after the handshake has completed.

Reply:
Very cool! Thank you for explaining this. Sounds like Early-Data is only for optimization purposes added to TLSv1.3 . Although less-secure, I can see why developers may want it enabled for: 

> "GET|HEAD|OPTIONS" which are the HTTP methods considered idempotent and safe. 

If app developers are using GET request to delete records or initiate payments, then they have more serious problems to worry about than this nginx router. :smile: